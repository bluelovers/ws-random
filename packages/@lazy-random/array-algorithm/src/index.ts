
export * from './swapAlgorithm';
export * from './array_rebase';
