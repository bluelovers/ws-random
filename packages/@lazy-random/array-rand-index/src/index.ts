import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
