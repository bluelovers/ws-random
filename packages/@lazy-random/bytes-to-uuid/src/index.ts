import { BYTE_TO_HEX_TO_LOWER_CASE } from '@lazy-random/shared-lib';

export function _createBytesToUuidFn(bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	return (buf: ArrayLike<number>, offset?: number) =>
	{
		let i = offset || 0;
		// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
		return ([
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
		]).join('');
	}
}

/**
 * @see https://github.com/kelektiv/node-uuid/blob/master/lib/bytesToUuid.js
 */
export function bytesToUuid(buf: ArrayLike<number>, offset?: number, bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	let i = offset || 0;
	// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
	return ([
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
	]).join('');
}

export default bytesToUuid
