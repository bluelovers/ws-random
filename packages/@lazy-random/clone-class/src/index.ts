/**
 * @todo support typescript
 */
export function getClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o;

	if (thisArgv instanceof RNGClass)
	{
		// @ts-ignore
		o = (thisArgv.__proto__.constructor)
	}
	else
	{
		o = RNGClass
	}

	return o
}

/**
 * @todo support typescript
 */
export function cloneClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o = getClass(RNGClass, thisArgv, ...argv)

	// @ts-ignore
	return new o(...argv)
}

export default cloneClass
