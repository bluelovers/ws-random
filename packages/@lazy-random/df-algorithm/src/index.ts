
export { dfBates } from './bates'

export { dfBernoulli } from './bernoulli'
export { dfBinomial } from './binomial'
export { dfExponential } from './exponential'
export { dfGeometric } from './geometric'

export { dfIrwinHall } from './irwin-hall'
export { dfLogNormal } from './log-normal'

export { dfNormal } from './normal'
export { dfPareto } from './pareto'
