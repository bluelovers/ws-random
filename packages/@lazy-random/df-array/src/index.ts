
export { dfArrayIndex } from './array-index'
export { dfArrayShuffle } from './array-shuffle'

export { dfArrayUnique } from './array-unique'
export { dfArrayFill } from './array-fill'

