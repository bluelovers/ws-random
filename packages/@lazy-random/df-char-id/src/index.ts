/**
 * Created by user on 2018/11/9/009.
 */

import UString from "uni-string";
import { ENUM_ALPHABET } from '@lazy-random/shared-lib';
import { expect } from '@lazy-random/expect';
import { floatToString } from '@lazy-num/float-to-string';
import { randIndex as _randIndex } from '@lazy-random/util-distributions';
import { IRNGLike } from '@lazy-random/rng-abstract';

export function dfCharID(random: IRNGLike, char?: ENUM_ALPHABET | string | Buffer | number, size?: number)
{
	if (typeof char === 'number')
	{
		if (typeof size === 'number')
		{
			char = floatToString(char)
		}
		else
		{
			[size, char] = [char, null];
		}
	}

	size = size || 8;
	//ow(size, ow.number.integer.gt(0));
	expect(size).integer.gt(0)

	if (!char)
	{
		char = ENUM_ALPHABET.DEFAULT
	}

	const ls = UString.create(char).split('');
	const len = ls.length;

	expect(ls).to.have.lengthOf.gt(1);

	const randIndex = () =>
	{
		return _randIndex(random, len)
	};

	return () =>
	{
		let i = size;
		let list: string[] = [];
		while (i--)
		{
			list.push(ls[randIndex()])
		}
		return list.join('');
	}
}

export default dfCharID
