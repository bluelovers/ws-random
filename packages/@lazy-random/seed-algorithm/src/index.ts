
export * from './algorithm/string/xfnv1a';
export * from './algorithm/string/xmur3';
export * from './algorithm/float/int_x_2';
export * from './algorithm/int-list/v3b';

export * from './seed/seed-list';
