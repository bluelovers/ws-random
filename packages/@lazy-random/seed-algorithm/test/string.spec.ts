import { df_xfnv1a, df_xfnv1a_2, df_xmur3, df_xmur3a } from '../src/index';

describe(`string`, () =>
{

	([
		df_xfnv1a,
		df_xfnv1a_2,
		df_xmur3,
		df_xmur3a,
	]).forEach(fn =>
	{

		test(fn.name, () =>
		{

			const seedFn: () => number = fn('apple')

			let actual = [
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
				seedFn(),
			];

			expect(actual).toMatchSnapshot();

			const seedFn2: () => number = fn('applE')

			let actual2 = [
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
				seedFn2(),
			];

			expect(actual).not.toStrictEqual(actual2);

			expect(actual2).toMatchSnapshot();

		});

	})

})
