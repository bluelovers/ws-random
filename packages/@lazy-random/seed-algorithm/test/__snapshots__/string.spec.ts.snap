// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string df_xfnv1a 1`] = `
[
  1880952902,
  3822283825,
  478043131,
  273559758,
  1523837525,
  2343075387,
  289972495,
  310198244,
  41357768,
  971392513,
  256763979,
  3324999188,
  2910291732,
  3994868516,
  1909282480,
  784123824,
]
`;

exports[`string df_xfnv1a 2`] = `
[
  3813144010,
  1070432286,
  792785484,
  1765074406,
  4015605075,
  334274345,
  1298197448,
  139970592,
  2429869922,
  3172324050,
  3345526116,
  118171613,
  1317279240,
  1554938902,
  2251740804,
  1844838313,
]
`;

exports[`string df_xfnv1a_2 1`] = `
[
  3759755561,
  1147060720,
  4097353528,
  1955590951,
  1754306856,
  795370096,
  663350128,
  3014854124,
  1482522165,
  248644456,
  362428993,
  652016822,
  2033568404,
  2498890281,
  2307762092,
  2694967605,
]
`;

exports[`string df_xfnv1a_2 2`] = `
[
  3757097379,
  347599353,
  494481804,
  417145549,
  2138963017,
  2342773473,
  3731513101,
  437964522,
  1184312788,
  690238350,
  2436099395,
  724923474,
  385365181,
  909001012,
  660328465,
  3554879605,
]
`;

exports[`string df_xmur3 1`] = `
[
  1463900127,
  3382348051,
  4017611752,
  4205503226,
  1279010286,
  2149068837,
  2433646083,
  1467709509,
  3541791720,
  1221864153,
  269156765,
  2490544737,
  1955113118,
  2746753104,
  1646923655,
  3647521537,
]
`;

exports[`string df_xmur3 2`] = `
[
  3071847068,
  3608824145,
  769630472,
  2919198978,
  4262893782,
  724545136,
  1193843432,
  1985298288,
  1627391293,
  1332188356,
  1051812219,
  2308652250,
  1818767607,
  3494459685,
  3067848249,
  370396522,
]
`;

exports[`string df_xmur3a 1`] = `
[
  3032347235,
  3966414462,
  272600060,
  1311513922,
  1740450750,
  3790462781,
  2367108256,
  3574728707,
  4067836055,
  1875714520,
  179541701,
  14514027,
  1832936131,
  3600805452,
  3037423323,
  3707002597,
]
`;

exports[`string df_xmur3a 2`] = `
[
  4243690504,
  1134027138,
  3456639521,
  2828821235,
  3349268413,
  3090301747,
  2432799623,
  2924086854,
  4145921600,
  1049769401,
  30172453,
  1246630121,
  3174435774,
  4266128259,
  36855377,
  1412169984,
]
`;
